include/robot_interfaces/robot_logger.hpp

```diff
@@ -0,0 +1,331 @@
+    Action desired_action;
+    Observation observation;
+    Status status;
+
+    std::shared_ptr<robot_interfaces::RobotData<Action, Observation, Status>>
+        logger_data_;
+
+    std::vector<std::string> header_;
+
+    int block_size_;
+    long int index_;
+
+    bool stop_was_called_;
+
+    std::ofstream output_file_;
+    std::string output_file_name_;
+
+    RobotLogger(std::shared_ptr<
+                    robot_interfaces::RobotData<Action, Observation, Status>>
+                    robot_data,
+                int block_size)
+        : block_size_(block_size),
+          stop_was_called_(false),
+          logger_data_(robot_data)
+    {
+        thread_ = std::make_shared<real_time_tools::RealTimeThread>();
+    }
+
+    virtual ~RobotLogger()
+    {
+        stop_was_called_ = true;
+        stop();
+        thread_->join();
+    }
+
+    std::vector<std::string> get_header()
+    {
+
+        std::vector<std::string> observation_name = observation.get_name();
+        std::vector<std::string> applied_action_name =
+            applied_action.get_name();
+        std::vector<std::string> desired_action_name =
+            desired_action.get_name();
+        std::vector<std::string> status_name = status.get_name();
+
+        std::vector<std::vector<double>> observation_data =
+            observation.get_data();
+        std::vector<std::vector<double>> applied_action_data =
+            applied_action.get_data();
+        std::vector<std::vector<double>> desired_action_data =
+            desired_action.get_data();
+        std::vector<std::vector<double>> status_data = status.get_data();
+
+        // using the newest index information in logger_data to extract
+        // information on the fields to be logged
+        applied_action =
+            (*logger_data_->applied_action)[logger_data_->applied_action
+                                                ->newest_timeindex()];
+        desired_action =
+            (*logger_data_->desired_action)[logger_data_->desired_action
+                                                ->newest_timeindex()];
+        observation =
+            (*logger_data_
+                  ->observation)[logger_data_->observation->newest_timeindex()];
+        status =
+            (*logger_data_->status)[logger_data_->status->newest_timeindex()];
+
+        header_.push_back("#");
+        header_.push_back("[Timestamp]");
+        header_.push_back("(Time Index)");
+
+        append_name_to_header(status_name, status_data);
+        append_name_to_header(observation_name, observation_data);
+        append_name_to_header(applied_action_name, applied_action_data);
+        append_name_to_header(desired_action_name, desired_action_data);
+
+        return header_;
+
+    }
+
+    void append_name_to_header(std::vector<std::string> field_name,
+                               std::vector<std::vector<double>> field_data)
+    {
+
+        int i, j;
+        std::string temp;
```

Comment:
The scope of those can be reduced even more.
- i and j could be defined directly in the loop header and should actually of type `size_t`: `for (size_t i = 0; i < ...`
- temp can be defined where it is assigned (`std::string temp = field_name[i] ...`)